### ISSUE TITLE
[Feature request] - Better Snap Points and Geometric analysis possible?

### ISSUE THREAD

**Issue by 0lm** [tannerhelland/PhotoDemon#498]
### Is this feature request related to a problem? 

Half half I would say. The feature would make alot of work very convenient

### Describe the best way to implement this.

Currently, Photo Demon's snap point system feels non-existend. Most of time, when I move objects, or want to add different images to a pre-created canvas size, I need to zoom in alot, to be even able to pull an image to the very corner of the canvas, without accidently pulling it too far away, so part of the image is out of the canvas

Is it possible to add a (better?) snap point system, where especially when you reach the canvas borders, it will hard snap immediately, needing you to pull again to overcome that hard snap? (similar to photoshop). And talking about snap points, is it possible to also add these geometric lines, which show you where the mid of the image is, or how far you are away from a certain object etc, while moving an object? So, I can, for example, snap added text boxes, or added images, in the mid of the canvas, or for example, two images, both same distance away from left and right canvas borders? Or two added images/text boxes on the same height level, or when resizing text boxes, to see, when it has the same size, by seeing these lines etc.

Im not sure how to describe it, but I hope you know how photoshop's snap system works. Then you will understand, I think. Im not sure if this is too much ask,, though. Not sure how hard/easy it is to implement a better/customizeable snap system.

EDIT: Just looked up some terms. I am referring to things like "snap to pixel", "snap to grid", or stuff like snapping to other objects on the canvas. But not sure how these distance measurment lines are called, though, when you move an object.

### Thank you!

If this feature makes it into PhotoDemon, I am happy to add your name to PhotoDemon's contributor list.  Please let me know what name (and optional website link) you'd like me to use.

**Comment by tannerhelland** [comment]
Great suggestion, @0lm.  Thank you.  I have added this to my short-term to-do list.

**Comment by tannerhelland** [comment]
Thank you for the follow-up, @0lm !  I'm sorry progress on this issue has been slow.

I am working on snapping [in a separate branch](https://github.com/tannerhelland/PhotoDemon/tree/snap-support) and snapping is now available when moving layers.  I haven't solved snapping for selections yet, and snapping for layer resizing also remains to be solved.  But progress is (slowly) being made.

I hope it won't be too much longer before snap features are merged to the main branch.  Thank you for continuing to track this issue.

**Comment by tannerhelland** [comment]
Snapping support is now available in nightly builds.  Thank you for your patience!

I will leave this issue open for a few days while I continue testing the feature.  I also have a few more small snap-related enhancements I would like to add.

Thank you again for suggesting this feature.  I hope it helps!

**Comment by tannerhelland** [comment]
Yes, exactly right!  As I type this, I am committing a "smart guides" feature and uploading a new nightly build.  (This is also in the `View` menu, and when enabled it will draw a highlight along the edge(s) being snapped.  It is enabled by default.)

I'm really happy with how this feature turned out, and hopefully future feature suggestions won't take me quite so long to implement 😅 Thank you for testing the snapping!
